File: web_BACKUP_BEFORE_NO_MULTIPLEX/src/lib/replay.before_ui/commentaryComposer.ts
import type {
  CommentaryAction,
  NarrativeIntent,
} from "./types";

function findPlayerOne(
  action: CommentaryAction,
  category: string,
): string | null {
  for (const event of action.sub_events) {
    if (event.category === category) {
      return event.player_one_name ?? null;
    }
  }

  return null;
}

function findPlayerTwo(
  action: CommentaryAction,
  category: string,
): string | null {
  for (const event of action.sub_events) {
    if (event.category === category) {
      return event.player_two_name ?? null;
    }
  }

  return null;
}

function findPlayerOneId(
  action: CommentaryAction,
  category: string,
): number | null {
  for (const event of action.sub_events) {
    if (event.category === category) {
      return event.player_one_id ?? null;
    }
  }

  return null;
}

function findPlayerTwoId(
  action: CommentaryAction,
  category: string,
): number | null {
  for (const event of action.sub_events) {
    if (event.category === category) {
      return event.player_two_id ?? null;
    }
  }

  return null;
}

export function composeAction(
  action: CommentaryAction,
): NarrativeIntent {
  const categories = action.sub_events.map(
    (event) => event.category,
  );

  let kind = "other";

  let creator: string | null = null;
  let shooter: string | null = null;
  let defender: string | null = null;
  let goalkeeper: string | null = null;
  let substituteOn: string | null = null;
  let substituteOff: string | null = null;

  if (categories.includes("GOAL")) {
    kind = "goal";

    creator = findPlayerOne(action, "CHANCE");
    shooter = findPlayerOne(action, "GOAL");

    if (shooter === null) {
      shooter = findPlayerOne(action, "SHOT");
    }

    if (shooter === null) {
      shooter = creator;
    }
  } else if (categories.includes("SAVE")) {
    kind = "chance_saved";

    creator = findPlayerOne(action, "ASSISTEDCHANCE");
    shooter = findPlayerTwo(action, "ASSISTEDCHANCE");

    if (shooter === null) {
      shooter = findPlayerOne(action, "SHOT");
    }

    if (shooter === null) {
      shooter = findPlayerOne(action, "CHANCE");
    }

    goalkeeper = findPlayerOne(action, "SAVE");
  } else if (categories.includes("OFFTARGET")) {
    kind = "chance_offtarget";

    creator = findPlayerOne(action, "ASSISTEDCHANCE");
    shooter = findPlayerTwo(action, "ASSISTEDCHANCE");

    if (shooter === null) {
      shooter = findPlayerOne(action, "SHOT");
    }

    if (shooter === null) {
      shooter = findPlayerOne(action, "CHANCE");
    }
  } else if (categories.includes("TACKLE")) {
    kind = "chance_tackled";

    creator = findPlayerOne(action, "ASSISTEDCHANCE");
    shooter = findPlayerTwo(action, "ASSISTEDCHANCE");

    defender = findPlayerOne(action, "TACKLE");

    if (shooter === null) {
      shooter = findPlayerOne(action, "CHANCE");
    }
  } else if (categories.includes("SUB")) {
    kind = "substitution";

    substituteOn = findPlayerOne(action, "SUB");
    substituteOff = findPlayerTwo(action, "SUB");
  } else if (categories.includes("SHOT")) {
    kind = "shot";
    shooter = findPlayerOne(action, "SHOT");
  }

  return {
    action_id: action.comm_event_id,
    minute: action.time,
    kind,

    creator_player: creator,
    shooter_player: shooter,
    defender_player: defender,
    goalkeeper,
    substitute_on: substituteOn,
    substitute_off: substituteOff,

    creator_player_id:
      findPlayerOneId(action, "ASSISTEDCHANCE") ??
      findPlayerOneId(action, "CHANCE"),

    shooter_player_id:
      findPlayerOneId(action, "GOAL") ??
      findPlayerTwoId(action, "ASSISTEDCHANCE") ??
      findPlayerOneId(action, "SHOT") ??
      findPlayerOneId(action, "CHANCE"),

    defender_player_id:
      findPlayerOneId(action, "TACKLE"),

    goalkeeper_id:
      findPlayerOneId(action, "SAVE"),

    substitute_on_id:
      findPlayerOneId(action, "SUB"),

    substitute_off_id:
      findPlayerTwoId(action, "SUB"),

    club_name: action.club_one_name,
  };
}

export function composeActions(
  actions: CommentaryAction[],
): NarrativeIntent[] {
  return actions.map(composeAction);
}

File: src/commentary_composer.py
from __future__ import annotations

from dataclasses import dataclass

from src.commentary_actions import CommentaryAction


@dataclass(frozen=True)
class NarrativeIntent:
    action_id: int
    minute: int
    kind: str

    # Ruoli narrativi espliciti.
    # creator_player = chi crea/serve l'azione
    # shooter_player = chi conclude
    # defender_player = chi interrompe con un tackle
    creator_player: str | None
    shooter_player: str | None
    defender_player: str | None
    goalkeeper: str | None
    substitute_on: str | None
    substitute_off: str | None

    creator_player_id: int | None
    shooter_player_id: int | None
    defender_player_id: int | None
    goalkeeper_id: int | None
    substitute_on_id: int | None
    substitute_off_id: int | None

    club_name: str | None


def _find_player_one(
    action: CommentaryAction,
    category: str,
) -> str | None:
    for event in action.sub_events:
        if event.category == category:
            return event.player_one_name
    return None


def _find_player_two(
    action: CommentaryAction,
    category: str,
) -> str | None:
    for event in action.sub_events:
        if event.category == category:
            return event.player_two_name
    return None


def _find_player_one_id(
    action: CommentaryAction,
    category: str,
) -> int | None:
    for event in action.sub_events:
        if event.category == category:
            return event.player_one_id
    return None


def _find_player_two_id(
    action: CommentaryAction,
    category: str,
) -> int | None:
    for event in action.sub_events:
        if event.category == category:
            return event.player_two_id
    return None


def compose_action(action: CommentaryAction) -> NarrativeIntent:
    categories = action.categories

    creator = None
    shooter = None
    defender = None
    goalkeeper = None
    substitute_on = None
    substitute_off = None

    if "GOAL" in categories:
        kind = "goal"

        creator = _find_player_one(action, "CHANCE")
        shooter = _find_player_one(action, "GOAL")

        if shooter is None:
            shooter = _find_player_one(action, "SHOT")

        if shooter is None:
            shooter = creator

    elif "SAVE" in categories:
        kind = "chance_saved"

        creator = _find_player_one(action, "ASSISTEDCHANCE")
        shooter = _find_player_two(action, "ASSISTEDCHANCE")

        if shooter is None:
            shooter = _find_player_one(action, "SHOT")

        if shooter is None:
            shooter = _find_player_one(action, "CHANCE")

        goalkeeper = _find_player_one(action, "SAVE")

    elif "OFFTARGET" in categories:
        kind = "chance_offtarget"

        creator = _find_player_one(action, "ASSISTEDCHANCE")
        shooter = _find_player_two(action, "ASSISTEDCHANCE")

        if shooter is None:
            shooter = _find_player_one(action, "SHOT")

        if shooter is None:
            shooter = _find_player_one(action, "CHANCE")

    elif "TACKLE" in categories:
        kind = "chance_tackled"

        creator = _find_player_one(action, "ASSISTEDCHANCE")
        shooter = _find_player_two(action, "ASSISTEDCHANCE")

        defender = _find_player_one(action, "TACKLE")

        if shooter is None:
            shooter = _find_player_one(action, "CHANCE")

    elif "SUB" in categories:
        kind = "substitution"

        # Nei dati ufficiali:
        # player_one = entrante
        # player_two = uscente
        substitute_on = _find_player_one(action, "SUB")
        substitute_off = _find_player_two(action, "SUB")

    elif "SHOT" in categories:
        kind = "shot"
        shooter = _find_player_one(action, "SHOT")

    else:
        kind = "other"

    return NarrativeIntent(
        action_id=action.comm_event_id,
        minute=action.time,
        kind=kind,
        creator_player=creator,
        shooter_player=shooter,
        defender_player=defender,
        goalkeeper=goalkeeper,
        substitute_on=substitute_on,
        substitute_off=substitute_off,

        creator_player_id=_find_player_one_id(action, "ASSISTEDCHANCE")
        or _find_player_one_id(action, "CHANCE"),

        shooter_player_id=_find_player_one_id(action, "GOAL")
        or _find_player_two_id(action, "ASSISTEDCHANCE")
        or _find_player_one_id(action, "SHOT")
        or _find_player_one_id(action, "CHANCE"),

        defender_player_id=_find_player_one_id(action, "TACKLE"),

        goalkeeper_id=_find_player_one_id(action, "SAVE"),

        substitute_on_id=_find_player_one_id(action, "SUB"),

        substitute_off_id=_find_player_two_id(action, "SUB"),

        club_name=action.club_one_name,
    )


def compose_actions(
    actions: tuple[CommentaryAction, ...],
) -> tuple[NarrativeIntent, ...]:
    return tuple(compose_action(action) for action in actions)
